File: platform/test/connection/openai-embeddings/fetch-httpx/test_case.py
#!/usr/bin/env python3
"""
Connection test for OpenAI Embeddings API via the internal fetch_httpx package.

Tests proxy, verify_ssl, ca_bundle, and client cert configuration
by resolving config from AppYamlConfig and making a health-check request
using fetch_httpx.AsyncClient directly.
"""

from __future__ import annotations

import argparse
import asyncio
import json
import os
import re
import sys
import time
from pathlib import Path

_HERE = Path(__file__).resolve().parent
_PLATFORM = _HERE
while _PLATFORM.name != "platform" and _PLATFORM != _PLATFORM.parent:
    _PLATFORM = _PLATFORM.parent


def _inject_paths():
    """Add polyglot/packages_py source dirs to sys.path from platform pyproject.toml."""
    toml = _PLATFORM / "pyproject.toml"
    if not toml.exists():
        return
    for line in toml.read_text().splitlines():
        m = re.search(r'path\s*=\s*"([^"]+)"', line)
        if not m:
            continue
        pkg_path = (_PLATFORM / m.group(1)).resolve()
        src = pkg_path / "src"
        sys.path.insert(0, str(src if src.is_dir() else pkg_path))


_inject_paths()

from app_yaml_load import load_app_yaml_config
from fetch_httpx import AsyncClient, Proxy


# ---------------------------------------------------------------------------
# Connection config resolution
# ---------------------------------------------------------------------------

def _resolve_connection_config(provider_cfg: dict) -> dict:
    """Resolve proxy, SSL, and cert settings from YAML + env."""
    return {
        "proxy_url": (
            provider_cfg.get("proxy_url")
            or os.environ.get("HTTPS_PROXY")
            or os.environ.get("HTTP_PROXY")
        ),
        "verify_ssl": provider_cfg.get("verify_ssl", True),
        "ca_bundle": (
            provider_cfg.get("ca_bundle")
            or os.environ.get("SSL_CERT_FILE")
            or os.environ.get("REQUESTS_CA_BUNDLE")
        ),
        "client_cert": provider_cfg.get("client_cert"),
        "client_key": provider_cfg.get("client_key"),
    }


def _resolve_api_key(provider_cfg: dict) -> str | None:
    """Resolve API key from config (with template guard) or env fallbacks."""
    cfg_key = provider_cfg.get("api_key") or ""
    if cfg_key and not cfg_key.startswith("{{"):
        return cfg_key
    return (
        os.environ.get("OPENAI_EMBEDDINGS_API_KEY")
        or os.environ.get("OPENAI_API_KEY")
    )


async def _test_connection(provider_cfg: dict, conn_cfg: dict) -> dict:
    """Test OpenAI Embeddings API connection using fetch_httpx.AsyncClient directly."""
    api_key = _resolve_api_key(provider_cfg)
    if not api_key:
        return {
            "skipped": True,
            "reason": "no OPENAI_EMBEDDINGS_API_KEY, OPENAI_API_KEY, or api_key in config",
        }

    base_url = provider_cfg.get("base_url") or "https://api.openai.com/v1"

    proxy_url = conn_cfg["proxy_url"]
    verify_ssl = conn_cfg["verify_ssl"]
    ca_bundle = conn_cfg["ca_bundle"]
    client_cert = conn_cfg["client_cert"]
    client_key = conn_cfg["client_key"]

    proxy = Proxy(url=proxy_url) if proxy_url else None
    verify = ca_bundle if ca_bundle else verify_ssl
    cert = (client_cert, client_key) if client_cert else None

    payload = {"model": "text-embedding-3-small", "input": "Hello, world!"}

    try:
        start = time.monotonic()
        async with AsyncClient(
            base_url=base_url,
            headers={
                "Authorization": f"Bearer {api_key}",
                "Content-Type": "application/json",
            },
            verify=verify,
            cert=cert,
            proxies=proxy,
            trust_env=True,
            timeout=15,
        ) as client:
            resp = await client.post("/embeddings", json=payload)
            latency = round((time.monotonic() - start) * 1000)
            data = resp.json() if resp.content else {}
            embedding = (
                data.get("data", [{}])[0].get("embedding", [])
                if isinstance(data, dict)
                else []
            )
            return {
                "skipped": False,
                "connected": resp.status_code < 400,
                "status_code": resp.status_code,
                "endpoint": "/embeddings",
                "has_embedding": len(embedding) > 0,
                "embedding_length": len(embedding),
                "latency_ms": latency,
                "connection": {
                    "base_url": base_url,
                    "proxy": proxy_url or "(none)",
                    "verify_ssl": verify_ssl,
                    "ca_bundle": ca_bundle or "(default)",
                    "has_client_cert": bool(client_cert),
                },
            }
    except Exception as exc:
        return {
            "skipped": False,
            "connected": False,
            "error": {"name": type(exc).__name__, "message": str(exc)},
        }


async def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--app-env", default="dev")
    parser.add_argument("--config-dir", required=True)
    args = parser.parse_args()

    app_cfg = load_app_yaml_config(config_dir=args.config_dir, app_env=args.app_env).config
    provider_cfg = app_cfg.get_nested("providers", "openai_embeddings", default={}) or {}
    conn_cfg = _resolve_connection_config(provider_cfg)

    connection_test = await _test_connection(provider_cfg, conn_cfg)

    summary = {
        "service": "openai-embeddings-fetch-httpx",
        "mode": "connection-test",
        "python_package": "fetch_httpx",
        "app_yaml_initialized": True,
        "connection_config": conn_cfg,
        "connection_test": connection_test,
        "overall": "PASS" if connection_test.get("connected") else (
            "SKIP" if connection_test.get("skipped") else "FAIL"
        ),
    }
    print(json.dumps(summary, indent=2, default=str))
    return 0


if __name__ == "__main__":
    raise SystemExit(asyncio.run(main()))

File: platform/test/connection/openai-embeddings/fetch-undici/test_case.mjs
#!/usr/bin/env node

/**
 * Connection test for OpenAI Embeddings API via the internal fetch_undici package.
 *
 * Tests proxy, verify_ssl, ca_bundle, and client cert configuration
 * by resolving config from AppYamlConfig and making a health-check request
 * using AsyncClient with TLSConfig and Proxy support.
 */

import fs from 'node:fs';
import path from 'node:path';
import { pathToFileURL } from 'node:url';

const args = process.argv.slice(2);
const readArg = (name, fallback) => {
  const idx = args.indexOf(name);
  return idx >= 0 ? args[idx + 1] : fallback;
};

const appEnv = readArg('--app-env', process.env.APP_ENV || 'dev');
const configDir = readArg('--config-dir', process.env.CONFIG_DIR);
if (!configDir) {
  console.error('Missing --config-dir');
  process.exit(2);
}

const platformRoot = path.resolve(configDir, '..', '..');
const importFile = async (absPath) => import(pathToFileURL(absPath).href);

const appYamlLoadMod = await importFile(path.join(platformRoot, 'polyglot/app_yaml_load/mjs/dist/index.js'));
const fetchUndiciMod = await importFile(path.join(platformRoot, 'packages_mjs/fetch_undici/dist/index.js'));

const { config: appConfig } = await appYamlLoadMod.loadAppYamlConfig({ configDir, appEnv });
const providerConfig = appConfig.getNested(['providers', 'openai_embeddings'], {}) || {};

// ---------------------------------------------------------------------------
// Connection config resolution
// ---------------------------------------------------------------------------

const connCfg = {
  proxyUrl: providerConfig.proxy_url || process.env.HTTPS_PROXY || process.env.HTTP_PROXY || null,
  verifySsl: providerConfig.verify_ssl ?? true,
  caBundlePath: providerConfig.ca_bundle || process.env.SSL_CERT_FILE || process.env.NODE_EXTRA_CA_CERTS || null,
  clientCert: providerConfig.client_cert || null,
  clientKey: providerConfig.client_key || null,
};

// ---------------------------------------------------------------------------
// API key resolution (template guard)
// ---------------------------------------------------------------------------

const cfgKey = providerConfig.api_key || '';
const apiKey = (cfgKey && !cfgKey.startsWith('{{'))
  ? cfgKey
  : (process.env.OPENAI_EMBEDDINGS_API_KEY || process.env.OPENAI_API_KEY || null);

const baseUrl = providerConfig.base_url || 'https://api.openai.com/v1';
const model = providerConfig.model || 'text-embedding-3-small';

// ---------------------------------------------------------------------------
// Connection test
// ---------------------------------------------------------------------------

let connectionTest;

if (!apiKey) {
  connectionTest = { skipped: true, reason: 'no OPENAI_EMBEDDINGS_API_KEY, OPENAI_API_KEY, or api_key in config' };
} else {
  try {
    const tlsConfig = fetchUndiciMod.createTLSConfig({
      verify: connCfg.verifySsl,
      ca: connCfg.caBundlePath ? fs.readFileSync(connCfg.caBundlePath) : null,
      cert: connCfg.clientCert ? fs.readFileSync(connCfg.clientCert) : null,
      key: connCfg.clientKey ? fs.readFileSync(connCfg.clientKey) : null,
    });

    const proxy = connCfg.proxyUrl ? fetchUndiciMod.createProxy(connCfg.proxyUrl) : null;

    const client = new fetchUndiciMod.AsyncClient({
      tls: tlsConfig,
      proxy,
      trustEnv: true,
    });

    const payload = JSON.stringify({ model, input: 'Hello, world!' });

    const start = performance.now();
    const response = await client.post(`${baseUrl}/embeddings`, {
      headers: {
        'Authorization': `Bearer ${apiKey}`,
        'Content-Type': 'application/json',
      },
      body: payload,
    });
    const latency = Math.round(performance.now() - start);

    const body = await response.json();
    await client.close();

    const embeddingLength = body.data?.[0]?.embedding?.length ?? 0;

    connectionTest = {
      skipped: false,
      connected: embeddingLength > 0,
      endpoint: `POST ${baseUrl}/embeddings`,
      status_code: response.statusCode,
      model: body.model,
      dimensions: embeddingLength,
      usage_total_tokens: body.usage?.total_tokens,
      latency_ms: latency,
      connection: {
        base_url: baseUrl,
        proxy: connCfg.proxyUrl || '(none)',
        verify_ssl: connCfg.verifySsl,
        ca_bundle: connCfg.caBundlePath || '(default)',
        has_client_cert: Boolean(connCfg.clientCert),
      },
    };
  } catch (err) {
    connectionTest = {
      skipped: false,
      connected: false,
      error: { name: err.name, message: err.message },
    };
  }
}

const summary = {
  service: 'openai-embeddings-fetch-undici',
  mode: 'connection-test',
  mjs_package: '@internal/fetch-undici',
  app_yaml_initialized: true,
  connection_config: connCfg,
  connection_test: connectionTest,
  overall: connectionTest.connected ? 'PASS' : connectionTest.skipped ? 'SKIP' : 'FAIL',
};

// Redact sensitive fields before logging (CodeQL js/clear-text-logging)
const redactedSummary = { ...summary, connection_config: { ...summary.connection_config, proxyUrl: summary.connection_config.proxyUrl ? '*******' : null, clientKey: summary.connection_config.clientKey ? '*******' : null } };
console.log(JSON.stringify(redactedSummary, null, 2));
